File: cpp/algo/src/structures/queue.cpp
#include "iostream"
#include "queue.h"
#include "../main.h"

using namespace std;

int peek(Queue<int> &queue) {
    try {
        return queue.peek();
    }
    catch (std::runtime_error &err) {
        cout << err.what() << endl;
        return -1;
    }
}

void queueExample() {
    cout << "Queue" << endl;

    Queue<int> queue;
    peek(queue);

    queue.enqueue(1);
    queue.enqueue(2);
    queue.enqueue(3);
    queue.enqueue(4);
    queue.enqueue(5);
    queue.enqueue(6);
    queue.enqueue(7);
    queue.enqueue(8);
    queue.enqueue(9);
    queue.enqueue(10);

    cout << peek(queue) << endl;

    try {
        int value;
        while ((value = queue.dequeue()) > -1) {
            cout << value << endl;
        }
    }
    catch (std::runtime_error& err) {
        cout << err.what() << endl;
    }

    cout << endl;
}

File: java/kata/src/main/java/br/com/ari/algo/structures/Queue.java
package br.com.ari.algo.structures;

import java.util.NoSuchElementException;

public class Queue<T> {

    private static class Node<T> {
        T value;
        Node<T> next;
    }

    private Node<T> head;
    private Node<T> tail;
    private int size;

    public void enqueue(T value) {
        Node<T> node = new Node<>();
        node.value = value;

        this.size++;

        if (this.tail == null) {
            this.head = this.tail = node;
            return;
        }

        this.tail.next = node;
        this.tail = node;
    }

    public T dequeue() {
        if (this.head == null) {
            throw new NoSuchElementException("Empty queue");
        }

        var node = this.head;
        this.head = this.head.next;
        this.size--;

        if (this.head == null) {
            this.tail = null;
        }

        return node.value;
    }

    public T peek() {
        if (this.head == null) {
            throw new NoSuchElementException("Empty queue");
        }

        return this.head.value;
    }

    public int size() {
        return size;
    }

}
